### PR TITLE
WAZO-2995 Group chat functionality

### PR DESCRIPTION
**Pull request by MehdiEmrani**
change the limitation of UserRoom Length from equal=2 to min=2,max=100,
add test_create_group test,
change test_create_with_wrong_users_number to pass the new validation rule,
change the _assert_create_raise_400_users_error to check the constraint rule

### CODE REVIEW

**Comment by fblackburn1** [comment]
To fix black linter issue you can run `tox -e black` then commit changes

**Review by fblackburn1 — COMMENTED** [review]
Thank you for the PR, it's nice to see new contributors :)
As discussed, here's a quick feedback
Removing `equal=2` is essentially what is needed to make group chat :+1:

Then a lot of questions about workflow need to be addressed to complete "the group chat feature"
ex: 
* How handle privacy when member is added to the conversation later
* How to define max participants
* How to leave a conversation
* Documentation
* etc

File: integration_tests/suite/test_user_room.py [hunk 1]
```diff
@@ -44,20 +44,31 @@ def test_list(self, room_1, room_2, _):
...
 
-    def test_create(self):
+    def _get_token_uuid(self):
+        return f'events.chatd.users.{TOKEN_USER_UUID}.rooms.created'
+
+    def _get_uuid_acl(self):
+        return f'events.chatd.users.{UUID}.rooms.created'
+
+    def _get_uuid_2_acl(self):
+        return f'events.chatd.users.{UUID_2}.rooms.created'
```
[the remark this answers is not in the document]
Comment (MehdiEmrani) [reply integration_tests/suite/test_user_room.py:67]: yes, I agree with you but the line length limitation of tox-linter forced me. tab indents of function and if block plus the length of the strings exceeded the limitation.

line_too_long_error

File: integration_tests/suite/test_user_room.py [hunk 2]
```diff
@@ -140,16 +197,24 @@ def test_create_with_wrong_users_number(self):
...
+                #   'uuid': str(TOKEN_USER_UUID),
+                #   'tenant_uuid': str(TOKEN_TENANT_UUID),
+                #   'wazo_uuid': str(WAZO_UUID)
+                # },
                 {'uuid': UUID, 'tenant_uuid': UUID, 'wazo_uuid': UUID},
                 {'uuid': UUID_2, 'tenant_uuid': UUID_2, 'wazo_uuid': UUID_2},
             ]
         }
+        room_args = self._add_user_range(room_args)
+
```
[the remark this answers is not in the document]
Comment (MehdiEmrani) [reply integration_tests/suite/test_user_room.py]: yes, I removed the first  `room_args`

**Comment by miladrazban** [comment]
> Thank you for the PR, it's nice to see new contributors :) As discussed, here's a quick feedback Removing `equal=2` is essentially what is needed to make group chat 👍
> 
> Then a lot of questions about workflow need to be addressed to complete "the group chat feature" ex:
> 
> * How handle privacy when member is added to the conversation later
> * How to define max participants
> * How to leave a conversation
> * Documentation
> * etc

Miker here, 
I'm glad that you mentioned these questions, I think in order to have a solid outcome we need to have a roadmap and define what is a group chat exactly, so that we are on the same page.

**Comment by sboily** [comment]
Hello, yes agreed. So would you like make a proposal and we could work together to construct the roadmap for this feature. A google doc would be a good start.

**Comment by miladrazban** [comment]
> Hello, yes agreed. So would you like make a proposal and we could work together to construct the roadmap for this feature. A google doc would be a good start.

I created a Google doc for the  group chat feature. Please use the link below:

https://docs.google.com/document/d/1j-U7rpwZaQahfRWngVU-YgQUZzislIzhksLn5zRERsQ/edit?usp=sharing

File: integration_tests/suite/test_user_room.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply integration_tests/suite/test_user_room.py]: Ah our friendly length limitation :laughing: 
You can see how it was solved [here](https://github.com/wazo-platform/wazo-chatd/blob/master/integration_tests/suite/test_room_message.py#L160-L182)
But at this point, it's merely a matter preference

**Comment by PSV37** [comment]
HI  @pc-m @sboily @miladrazban @fblackburn1  

`/opt/venv/lib/python3.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
Could not read config file /var/lib/wazo-auth-keys/wazo-chatd-key.yml: [Errno 2] No such file or directory: '/var/lib/wazo-auth-keys/wazo-chatd-key.yml'
2022-03-30 18:25:02,871 [1] (ERROR) (wazo_chatd.main): undefined environment variable XIVO_UUID
2022-03-30 18:25:02,871 [1] (CRITICAL) (root): XIVO_UUID environment variable is not set
Traceback (most recent call last):
  File "/opt/venv/bin/wazo-chatd", line 11, in <module>
    load_entry_point('wazo-chatd==1.0', 'console_scripts', 'wazo-chatd')()
  File "/opt/venv/lib/python3.7/site-packages/wazo_chatd-1.0-py3.7.egg/wazo_chatd/main.py", line 30, in main
    set_xivo_uuid(conf, logger)
  File "/opt/venv/lib/python3.7/site-packages/xivo/config_helper.py", line 148, in set_xivo_uuid
    config['uuid'] = get_xivo_uuid(logger)
  File "/opt/venv/lib/python3.7/site-packages/xivo/config_helper.py", line 143, in get_xivo_uuid
    raise UUIDNotFound()
xivo.config_helper.UUIDNotFound: XIVO_UUID environment variable is not set
`
how can we get environment variables 


Thank you

**Comment by pc-m** [comment]
> HI @pc-m @sboily @miladrazban @fblackburn1
> 
> `/opt/venv/lib/python3.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>. """) Could not read config file /var/lib/wazo-auth-keys/wazo-chatd-key.yml: [Errno 2] No such file or directory: '/var/lib/wazo-auth-keys/wazo-chatd-key.yml' 2022-03-30 18:25:02,871 [1] (ERROR) (wazo_chatd.main): undefined environment variable XIVO_UUID 2022-03-30 18:25:02,871 [1] (CRITICAL) (root): XIVO_UUID environment variable is not set Traceback (most recent call last): File "/opt/venv/bin/wazo-chatd", line 11, in <module> load_entry_point('wazo-chatd==1.0', 'console_scripts', 'wazo-chatd')() File "/opt/venv/lib/python3.7/site-packages/wazo_chatd-1.0-py3.7.egg/wazo_chatd/main.py", line 30, in main set_xivo_uuid(conf, logger) File "/opt/venv/lib/python3.7/site-packages/xivo/config_helper.py", line 148, in set_xivo_uuid config['uuid'] = get_xivo_uuid(logger) File "/opt/venv/lib/python3.7/site-packages/xivo/config_helper.py", line 143, in get_xivo_uuid raise UUIDNotFound() xivo.config_helper.UUIDNotFound: XIVO_UUID environment variable is not set `
> 
> how can we get environment variables
> 
> Thank you

It depends on the context. On a Wazo stack this environment variable should always be set. If you're using docker you should use `--env` or `--env-file`

**Comment by sboily** [comment]
Hello, what is the status for this PR. We also worked on group chat on our latest hackathon. Honestly i think it's good idea to have this feature inside the project. @fblackburn1 what's your recommandation for this PR.

**Comment by fblackburn1** [comment]
Hey @miladrazban @MehdiEmrani Sorry for this (very) long delay
I rebased/updated and improved the branch to merge it
It will be available on the next release (23.01)
Thank you again for your contribution :pray:

**Comment by miladrazban** [comment]
> Hey @miladrazban @MehdiEmrani Sorry for this (very) long delay I rebased/updated and improved the branch to merge it It will be available on the next release (23.01) Thank you again for your contribution 🙏

Hey @fblackburn1  thanks for the update.
Quick question, how does the group chat management will look like? (such as add/remove users, etc)

**Comment by MehdiEmrani** [comment]
> Hey @miladrazban @MehdiEmrani Sorry for this (very) long delay I rebased/updated and improved the branch to merge it It will be available on the next release (23.01) Thank you again for your contribution 🙏

It's would be great to accomplish this. Thank you

**Comment by fblackburn1** [comment]
> Hey @fblackburn1 thanks for the update. Quick question, how does the group chat management will look like? (such as add/remove users, etc)

For now we didn't think about it and it's not on active development
This PR has been added as it, but same restriction apply from 1:1 chat

Feel free to make proposition (with exploration, comparison, pros/cons, etc..) through the gdrive above or github issue, before any implementation/PR

**Pull request merged**